docs: add recipient schemas (schema management) (#1520)
knocklabs/docs: content/managing-recipients/schemas.mdx

Curating a schema shapes how recipient properties appear across the Knock dashboard: in the template editor, in condition and segment builders, and in recipient tables. It also makes your data legible to the Knock agent, which reads your schema to understand what properties exist and what they mean.

## How recipient schemas work

As you [identify recipients](/managing-recipients/identifying-recipients), Knock observes the properties on that data and records them on the schema for that recipient type. Each property carries:

- **Key.** The property name, such as `plan` or `address.city`.
- **Type.** The data type Knock inferred from your data, such as `string`, `number`, `boolean`, `object`, or `array`. Types are read-only.
- **Source.** Whether Knock inferred the property from your data or it is a built-in system property, such as `email` on a user. This is read-only.
- **Preview text.** A value Knock has seen for the property, used to render realistic previews of templates that reference it.

> Schemas are generated per [environment](/version-control/environments), the same way recipient data is. A property that only appears in production will not show up on your development schema until Knock sees it there.

Reviewer: ```suggestion
Schemas are logically isolated per [environment](/version-control/environments), the same way recipient data is. A property that only appears in production will not show up on your development schema until Knock sees it there.
```
Open to other phrasings here, but the current sentence structure implies that recipient data is generated.

<Callout
  type="info"

>   title="Recipient schemas cover recipients"

Reviewer: ```suggestion
  title="Recipient schemas cover recipients."
```
nit per style guide

  text={
    <>
      Schema management applies to recipient data: users, tenants, and object
      collections. Source events, workflow trigger data, and broadcast data are
      not part of recipient schemas.
    </>
  }
/>

## Curate a schema in the dashboard

> Go to **Settings** > **Schemas** and select a recipient type: users, tenants, or one of your object collections. You will see every property Knock has inferred for that type, along with its type, source, and preview text.

Reviewer: ```suggestion
Go to <a href="https://dashboard.knock.app/~/settings/schemas" target="_blank">**Settings** > **Schemas**</a> and select a recipient type: users, tenants, or one of your object collections. You will see every property Knock has inferred for that type, along with its type, source, and preview text.
```

### Hide or show a property

> Toggle a property's visibility to control where Knock offers it. When you hide a property, Knock stops offering it across the dashboard: template variable suggestions, condition and segment builders, and recipient table columns no longer list it.

Reviewer: ```suggestion
Toggle a property's visibility to control where Knock offers it. When you hide a property, Knock stops offering it across the dashboard: template variable suggestions, condition and audience segment builders, and recipient table columns no longer list it.
```
Maybe redundant, added for clarity about which feature it refers to

<Callout
  type="info"

>   title="Hiding never breaks anything"

Reviewer: ```suggestion
  title="Hiding never breaks anything."
```
style guide nit

## Where curated schemas appear

Once you curate a schema, it drives every place Knock surfaces recipient properties, including:

- **Template editor.** Variable suggestions offer only visible properties, and previews use your preview text.
- **Condition and segment builders.** Property pickers for [conditions](/designing-workflows/step-conditions) and [audiences](/concepts/audiences) list only visible properties.
- **Recipient tables.** Column options reflect the curated schema.

## Permissions

Access to schema management is governed by three permissions, each tied to a [role](/manage-your-account/roles-and-permissions).

| Capability                                          | Permission            | Roles                |
| --------------------------------------------------- | --------------------- | -------------------- |
| View schemas                                        | `item_schemas:read`   | Member, Admin, Owner |
| Set the preview text, or add a property             | `item_schemas:edit`   | Member, Admin, Owner |

> | Hide or show a property, add a label or description | `item_schemas:manage` | Admin, Owner         |

Reviewer: Wondering if the **Permission** column is actually helpful here? Maybe in the future once we have granular controls for these on custom roles/permissions? Just wondering if these values are meaningful to a customer right now.

Author: Agreed, removed the Permission column. The slugs aren't surfaced anywhere else in the docs, so the table now just maps capabilities to roles.

## Environments and branches

> Schema configuration is scoped to an environment and applies immediately, the same way [environment variables](/concepts/variables) do, rather than through commit and promote. Curate the schema in the environment whose data you want to manage.

Reviewer: ```suggestion
Schema configuration is scoped to an environment and applies immediately when saved, the same way [environment variables](/concepts/variables) do, rather than through commit and promote. Curate the schema in the environment whose data you want to manage.
```

Each schema is a list of properties:

```json title="A user schema"
{
  "item_type": "user",
  "item_id": null,
  "properties": [
    {
      "key": "plan",
      "type": "string",
      "preview_text": "\"enterprise\"",
      "visible": true,
      "label": "Plan",
      "description": "The account's current billing plan."
    }
  ]
}
```

The curatable fields on a property are:

| Field          | Type    | Description                                                               |
| -------------- | ------- | ------------------------------------------------------------------------- |
| `visible`      | boolean | Whether the property is offered across the dashboard. Defaults to `true`. |
| `label`        | string  | An optional human-readable display name.                                  |
| `description`  | string  | An optional description of the property.                                  |
| `preview_text` | string  | The preview text, JSON-encoded, used to preview templates.                |

A property's `type` and `source` are inferred by Knock and are read-only.

<Callout
  type="info"

>   title="Permissions apply to the Management API too"

Reviewer: ```suggestion
  title="Permissions apply to the Management API too."
```
style guide nit

  text={
    <>
      Reads require <code>item_schemas:read</code>. Changing a property's
      preview text or adding a property requires <code>item_schemas:edit</code>.
      Hiding or showing a property, or setting a label or description, requires{" "}

>       <code>item_schemas:manage</code>.

Reviewer: Same question here about the actual permissions, would natural-language descriptions be better?

Author: Yes, rewrote this callout in natural language tied to roles instead of permission slugs.

    </>
  }
/>

## Sync schemas with the CLI

The [Knock CLI](/cli/overview) can pull schemas from an environment to local files and push local changes back, so you can keep your schema configuration in version control alongside your other Knock resources.

```bash title="Pull and push schemas"
# Pull every schema in the environment into a local schemas directory
knock schema pull --all

# Push a single recipient schema to an environment

> knock schema push user --environment production

Reviewer: ```suggestion
knock schema push user --environment=production
```
Should this and similar flag usage include an `=` for the value? That's how we use flags in other commands

Author: Yes, switched to `=` here and in the CLI reference examples.
